pkg/apis/metrics/types.go

```diff
@@ -16,14 +16,205 @@ limitations under the License.
+	Containers []RawContainerMetrics `json:"containers" patchStrategy:"merge" patchMergeKey:"name"`
+	// Historical metric samples of pod-level resources.
+	Samples []PodSample `json:"samples" patchStrategy:"merge" patchMergeKey:"sampleTime"`
+}
+
+// RawPodMetricsList holds a list of RawPodMetrics.
+// Endpoints which return this type respect unversioned.ListOptions
+type RawPodMetricsList struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds
+	unversioned.ListMeta `json:"metadata,omitempty"`
+	// List of raw pod metrics.
+	Items []RawPodMetrics `json:"items"`
+}
+
+// RawContainerMetrics holds container-level unprocessed sample metrics.
+type RawContainerMetrics struct {
+	// Reference to the measured container.
+	Name string `json:"name"`
+	// Metadata labels associated with this container (not Kubernetes labels).
+	// For example, docker labels.
+	Labels map[string]string `json:"labels,omitempty"`
+	// Historical metric samples gathered from the container.
+	Samples []ContainerSample `json:"samples" patchStrategy:"merge" patchMergeKey:"sampleTime"`
+	// Application-specific metrics.
+	CustomMetrics []CustomMetric `json:"customMetrics,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+}
+
+// NonLocalObjectReference contains enough information to locate the referenced object.
+type NonLocalObjectReference struct {
+	Name      string    `json:"name"`
+	Namespace string    `json:"namespace"`
+	UID       types.UID `json:"uid,omitempty"`
+}
+
+// Sample defines metadata common to all sample types.
+// Samples may not be nested within other samples.
+type Sample struct {
+	// The time this data point was collected at.
+	SampleTime unversioned.Time `json:"sampleTime"`
+}
+
+// AggregateSample contains a metric sample point of data aggregated across containers.
+type AggregateSample struct {
+	Sample `json:",inline"`
+	// Metrics pertaining to CPU resources.
+	CPU *CPUMetrics `json:"cpu,omitempty"`
+	// Metrics pertaining to memory (RAM) resources.
+	Memory *MemoryMetrics `json:"memory,omitempty"`
+	// Metrics pertaining to network resources.
+	Network *NetworkMetrics `json:"network,omitempty"`
+}
+
+// PodSample contains a metric sample point of pod-level resources.
+type PodSample struct {
+	Sample `json:",inline"`
+	// Metrics pertaining to network resources.
+	Network *NetworkMetrics `json:"network,omitempty"`
+}
+
+// ContainerSample contains a metric sample point of container-level resources.
+type ContainerSample struct {
+	Sample `json:",inline"`
+	// Metrics pertaining to CPU resources.
+	CPU *CPUMetrics `json:"cpu,omitempty"`
+	// Metrics pertaining to memory (RAM) resources.
+	Memory *MemoryMetrics `json:"memory,omitempty"`
+}
+
+// NetworkMetrics contains data about network resources.
+type NetworkMetrics struct {
+	// Cumulative count of bytes received.
+	RxBytes *resource.Quantity `json:"rxBytes,omitempty"`
```

Comment:
Interesting note about converting to strings for values > 53 bits. So that's the reason behind Quantity - never knew that. I do like the idea of a separate repo with multiple language implementations - this has uses outside of k8s.